src/seasidecontactbuilder.cpp

```diff
@@ -0,0 +1,666 @@
+        contact.removeDetail(&nameDetail);
+        if (contact.details<QContactNickname>().isEmpty()) {
+            QString label = contact.detail<QContactDisplayLabel>().label();
+            if (label.isEmpty()) {
+                label = SeasideCache::generateDisplayLabelFromNonNameDetails(contact);
+            }
+            setNickname(contact, label);
+        }
+    }
+}
+
+/*
+ * Returns the index into the \a importedContacts list at which a
+ * duplicate (merge candidate) of the \a contact at the current
+ * \a contactIndex may be found, or \c -1 if no match is found.
+ *
+ * The default implementation uses a combination of GUID and
+ * name / label matching to determine if a contact is duplicated
+ * within the import list.
+ *
+ * The result will be used to merge any duplicated contacts within
+ * the import list, which is the required behaviour when performing
+ * an "import" style sync.  Any implementation which requires
+ * a one-to-one mapping between import contacts and stored device
+ * contacts should instead return -1 from this function.
+ */
+int SeasideContactBuilder::previousDuplicateIndex(QList<QContact> &importedContacts, QContact &contact, int contactIndex)
+{
+    const QString guid = contact.detail<QContactGuid>().guid();
+    const QString name = contactNameString(contact);
+    const bool emptyName = name.isEmpty();
+    const QString label = contact.detail<QContactDisplayLabel>().label().isEmpty()
+                        ?SeasideCache::generateDisplayLabelFromNonNameDetails(contact)
+                        : contact.detail<QContactDisplayLabel>().label();
+
+    int previousIndex = -1;
+    QHash<QString, int>::const_iterator git = d->importGuids.find(guid);
+    if (git != d->importGuids.end()) {
+        previousIndex = git.value();
+
+        if (!emptyName) {
+            // If we have a GUID match, but names differ, ignore the match
+            const QContact &previous(importedContacts[previousIndex]);
+            const QString previousName = contactNameString(previous);
+            if (!previousName.isEmpty() && (previousName != name)) {
+                previousIndex = -1;
+
+                // Remove the conflicting GUID from this contact
+                QContactGuid guidDetail = contact.detail<QContactGuid>();
+                contact.removeDetail(&guidDetail);
+            }
+        }
+    }
+    if (previousIndex == -1) {
+        if (!emptyName) {
+            QHash<QString, int>::const_iterator nit = d->importNames.find(name);
+            if (nit != d->importNames.end()) {
+                previousIndex = nit.value();
+            }
+        } else if (!label.isEmpty()) {
+            // Only if name is empty, use displayLabel - probably SIM import
+            QHash<QString, int>::const_iterator lit = d->importLabels.find(label);
+            if (lit != d->importLabels.end()) {
+                previousIndex = lit.value();
+            }
+        }
+    }
+
+    if (previousIndex == -1) {
```

Comment:
I would make this part a separate function, it is at odds with the expectations created by the function name.